backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/inoutcandidate/ShipmentScheduleRepository.java

```diff
@@ -416,77 +454,8 @@ public List<ShipmentSchedule> loadByPackageId(final @NonNull PackageId packageId
 				.collect(Collectors.toList());
 	}
 
-	public List<ShipmentScheduleId> listIdsByQuery(@NonNull final ShipmentScheduleQuery query)
+	public ImmutableSet<ShipmentScheduleId> getIdsByQuery(@NonNull final ShipmentScheduleQuery query)
 	{
-		if (query.getProductId() == null &&
-				query.getWarehouseId() == null &&
-				query.getAttributesKey() == null &&
-				query.getOrgId() == null)
-		{
-			throw new AdempiereException("At least one of the following parameters must be set: productId, warehouseId, attributesKey, orgId");
-		}
-		final IQueryBuilder<I_M_ShipmentSchedule> queryBuilder = queryBL.createQueryBuilder(I_M_ShipmentSchedule.class)
-				.addOnlyActiveRecordsFilter();
-		if (query.getProductId() != null)
-		{
-			queryBuilder.addEqualsFilter(I_M_ShipmentSchedule.COLUMNNAME_M_Product_ID, query.getProductId());
-		}
-		if (query.getWarehouseId() != null)
-		{
-			queryBuilder.addEqualsFilter(I_M_ShipmentSchedule.COLUMNNAME_M_Warehouse_ID, query.getWarehouseId());
-		}
-		if (query.getAttributesKey() != null)
-		{
-			queryBuilder.addEqualsFilter(I_M_ShipmentSchedule.COLUMNNAME_M_AttributeSetInstance_ID, query.getAttributesKey().getAsString(), ASIQueryFilterModifier.instance);
-		}
-		if (query.getOrgId() != null)
-		{
-			queryBuilder.addEqualsFilter(I_M_ShipmentSchedule.COLUMNNAME_AD_Org_ID, query.getOrgId());
-		}
-		if (query.isOnlyNonZeroQty())
-		{
-			queryBuilder.addNotEqualsFilter(I_M_ShipmentSchedule.COLUMNNAME_QtyReserved, 0);
-		}
-		return queryBuilder.create().listIds(ShipmentScheduleId::ofRepoId);
-	}
-
-	@Value
-	@Builder
-	public static class ShipmentScheduleQuery
-	{
-		@NonNull
-		@Builder.Default
-		QueryLimit limit = QueryLimit.NO_LIMIT;
-
-		Instant canBeExportedFrom;
-
-		APIExportStatus exportStatus;
-
-		@Builder.Default
-		boolean includeWithQtyToDeliverZero = false;
-
-		@Builder.Default
-		boolean includeInvalid = false;
-
-		@Builder.Default
-		boolean includeProcessed = false;
-
-		@Builder.Default
-		boolean fromCompleteOrderOrNullOrder = false;
-
-		@Builder.Default
-		boolean orderByOrderId = false;
-
-		/**
-		 * Only export a shipment schedule if its order does not have any schedule which is not yet ready to be exported.
-		 */
-		@Builder.Default
-		boolean onlyIfAllFromOrderExportable = false;
-
-		@Nullable OrgId orgId;
-		@Nullable AttributesKey attributesKey;
-		@Nullable ProductId productId;
-		@Nullable WarehouseId warehouseId;
-		@Builder.Default boolean onlyNonZeroQty = false;
+		return getMapBy(query).keySet();
```

Comment:
suggestion:
* refactor getMapBy and introduce: IQuery toSqlQuery(@NonNull final ShipmentScheduleQuery query)
* here you coud write toSqlQuery(query).listIds(...)